```{r} 
library(tidyverse) 
methylation_file_UL = 'D:/Birkbeck/Project/Ultralong/20_per_read_modified_base_calls.txt'
allele_file_UL = 'D:/BirkBeck/Project/Ultralong/20_per_read_variant_calls.txt'
positions <- c(7980, 10) # positions to define tentative haplotype 

chromosome_name = "KY962518.1_looped_2120_20_units"
coding_regions_h <- tribble(
  ~region,~start,~end,
  '18S',3658,5526,
  '5.8S',6597,6753,
  '28S',7921,12971)
```

```{r}
adjust_pos <- function(pos, loop_size=2120) {
  if_else(pos<=loop_size, pos-loop_size-1, pos-loop_size)
}

meth_calls_UL <- read_tsv(methylation_file_UL, col_types=cols()) %>%
  filter(chrm==chromosome_name) %>%
  select(-chrm) %>%
  mutate(pos=if_else(strand=='+', pos+1, pos),
         unit_num=as.integer( ( pos - 1) / 44838) + 1,
         meth_ref=adjust_pos((pos %% 44838), 2120)) %>%
  group_by(read_id, strand, can_log_prob, mod_log_prob, meth_ref) %>%
  summarise(unit_num=min(unit_num), .groups='drop') %>%
  select(read_id, meth_ref, can_prob=can_log_prob,
         mod_prob=mod_log_prob, unit_num) %>%
  pivot_longer(c(can_prob, mod_prob), values_to='prob', names_to='dominant') %>%
  filter(prob>log(0.9)) %>%
  group_by(read_id, meth_ref, unit_num) %>%
  distinct() %>%
  filter(n()==1) %>%
  mutate(meth=if_else(dominant=="mod_prob", 1, 0)) 
```

```{r}
allele_calls_UL <- read_tsv(allele_file_UL, col_types=cols()) %>%
  filter(chrm==chromosome_name) %>%
  select(-chrm) %>%
  mutate(pos=pos+1,
         unit_num=as.integer( ( pos - 1) / 44838) + 1,
         pos_ref=adjust_pos((pos %% 44838), 2120)) %>%
  group_by(read_id, strand, ref_log_prob, alt_log_prob,
           ref_seq, alt_seq, var_id, pos_ref) %>%
  summarise(unit_num=min(unit_num), .groups='drop') %>%
  select(read_id, pos_ref, ref=ref_seq, ref_prob=ref_log_prob,
         alt=alt_seq, alt_prob=alt_log_prob, unit_num) %>%
  pivot_longer(c(ref_prob, alt_prob), values_to='prob', names_to='dominant') %>%
  filter(prob>log(0.9)) %>%
  group_by(read_id, pos_ref, ref, alt, unit_num) %>%
  distinct() %>%
  filter(n()==1) %>%
  mutate(dominant=gsub('_prob', '', dominant),
         allele=if_else(dominant=="ref", ref, alt)) %>%
  ungroup() %>%
  select(read_id, pos_ref, allele, unit_num) %>%
  distinct() %>%
  group_by(read_id, unit_num) %>%
  filter(n_distinct(pos_ref)==n()) %>%
  group_by(pos_ref) %>%
  filter(n_distinct(allele) > 1) %>%
  mutate(allele_num=dense_rank(allele)) %>%
  ungroup()
```

```{r}
coding_reads_UL <- meth_calls_UL %>%
  group_by(read_id, unit_num) %>%
  summarise(mean_meth=mean(meth)) %>%
  inner_join(allele_calls_UL, by=c('read_id','unit_num')) %>%
  group_by(read_id, unit_num) %>%
  filter(! all(pos_ref > 0), ! all(pos_ref < 7980)) %>%
  group_by(read_id) %>%
  filter((pos_ref >= coding_regions_h$start[1] & pos_ref <= coding_regions_h$end[1]) | 
           (pos_ref >= coding_regions_h$start[2] & pos_ref <= coding_regions_h$end[2]) |
           (pos_ref >= coding_regions_h$start[3] & pos_ref <= coding_regions_h$end[3])) %>%
  group_by(pos_ref, allele) %>%
  summarise(num_reads=n()) %>%
  mutate(read_freq=num_reads/sum(num_reads))
ggplot(coding_reads_UL) + 
  geom_col(aes(x=as.factor(pos_ref), y=read_freq, fill=allele)) + 
  theme_bw() +
  labs(x='Position', y='Read Frequency') +
  theme(panel.border=element_rect(size=1.1),
        axis.text.y=element_text(size=7.5))
```
```{r}
title <- paste("Positions", paste(positions, collapse = "_"), sep="_")
filename <- paste0(title, paste(".png"))

#count of allele reads at position 7980
variants_test_UL <- allele_calls_UL %>% 
  group_by(read_id, unit_num) %>% 
  filter(pos_ref %in% positions) %>% 
  #filter(n() == 2) %>% 
  arrange(pos_ref) %>%
  pivot_wider(-allele_num, names_from = pos_ref, values_from = allele) %>%
  unite(haplotype, -read_id, -unit_num, sep='')
ggplot(variants_test_UL) + geom_bar(aes(x=haplotype))
```

```{r}
#Plot of tentative haplotype defined by the positions chosen in the first chunk
#x - position in unit
#y - proportion of reads with corresponding allele per tentative haplotype
variants_all_UL <- allele_calls_UL %>%
  select(read_id, pos_ref, allele) %>%
  distinct() %>%
  group_by(pos_ref) %>%
  filter(n_distinct(allele) > 1) %>%
  mutate(allele_num=dense_rank(allele)) %>%
  ungroup()

variants_join_UL <- variants_all_UL %>% 
  inner_join(variants_test_UL %>% group_by(haplotype) %>% filter(n()>10), by=c('read_id')) %>%
  group_by(pos_ref, haplotype, allele) %>%
  summarise(num_reads=n()) %>%
  group_by(pos_ref, haplotype) %>%
  mutate(read_freq=num_reads/sum(num_reads)) %>%
  filter(pos_ref >= 7921 & pos_ref <= 12971)

variants_join_UL %>% ggplot() + 
  geom_col(aes(x=as.factor(pos_ref), y=read_freq, fill=allele)) + 
  facet_grid(rows = vars(haplotype)) + 
  theme_bw() +
  labs(x='Position', y='Read Frequency') +
  theme(panel.grid=element_blank(),
        panel.border=element_rect(size=1.1),
        axis.text.x=element_text(size=8, angle=45, hjust=1 ))+
  ggtitle(title) 
```

```{r}
#Boxplot of methylation levels per allele at position 7980
methylation_UL <- meth_calls_UL %>%
  group_by(read_id, unit_num) %>%
  inner_join(allele_calls_UL, by=c('read_id','unit_num')) %>%
  group_by(read_id, unit_num) %>%
  filter(! all(pos_ref > 0), ! all(pos_ref < 7980)) %>%
  group_by(read_id)

methylation_UL %>% 
  filter(pos_ref == 7980) %>%
  filter(meth_ref > -1000, meth_ref <= 13332) %>%
  group_by(meth_ref, allele) %>%
    filter(n()>10) %>%
  summarise(meth=mean(meth), .groups='drop') %>%
  ggplot(aes(x=allele, y=meth)) +
  geom_boxplot(outlier.alpha = 0) + 
  geom_jitter(width=0.3) + 
  ggtitle("Ultralong 7980 methylation")
```